### DISCUSSION TITLE
Native mode, but packaged/frozen

### DISCUSSION THREAD

**Discussion by FabianGoessling** [zauberzeug/nicegui#1714] [Q&A]
### Question

Hey,
prior to opening an issue i wanted to ask my question here, as it might be a simple error of mine. I am packaging a
an application with nicegui frontend using the `shiv` package. This is essentially the same as pyinstaller, i.e. zipping the site-packages of the venv, so maybe someone has an idea how to solve my problem. 

As the problem only occurs after packaging, there is no easy way to reproduce the issue (needs shiv configs and so on..), so i try my best to explain. I have code as shown below with additional pywebview arguments, e.g. 'confirm_close', and a function to be run 'on_startup'.
If I execute the below code directly from my IDE this works, but after packaging the application the *window_args* and *on_startup* are not used. I dug into the nicegui source code to see what reaches the window_open() function and confirmed that the *window_args* reaching the pywebview call are empty. 

```python
from nicegui import ui, app

app.native.window_args['frameless'] = True #--> does not work
app.native.window_args['confirm_close'] = True #--> does not work
app.on_startup(do_some())

if __name__ == "__main__":
       ui.run(frameless=True) # --> works
```

I suppose this is something caused by the multiprocessing setup, but i am kind of stuck there..

Thanks for any ideas :)
Fabian

**Comment by rodja** [comment]
That is strange. Have you tried the same app with pyinstaller? Does the problem also occurs there?

**Comment by FabianGoessling** [comment]
I had some time to look into the problem in some more detail, but did not try with pyinstaller yet. Adjusting the  *open_window* function 
```python

def open_window(
    host: str, port: int, title: str, width: int, height: int, fullscreen: bool, frameless: bool,
    method_queue: mp.Queue, response_queue: mp.Queue, **kwargs
) -> None:
    while not helpers.is_port_open(host, port):
        time.sleep(0.1)

    window_kwargs = {
        'url': f'http://{host}:{port}',
        'title': title,
        'width': width,
        'height': height,
        'fullscreen': fullscreen,
        'frameless': frameless,
        **kwargs
        # **globals.app.native.window_args, <- remove
    }
    window = webview.create_window(**window_kwargs)
    closed = Event()
    window.events.closed += closed.set
    start_window_method_executor(window, method_queue, response_queue, closed)
    webview.start(storage_path=tempfile.mkdtemp(), **globals.app.native.start_args)

```
and the mp Process call in the *activate* function to 
```python

def activate(host: str, port: int, title: str, width: int, height: int, fullscreen: bool, frameless: bool) -> None:
    def check_shutdown() -> None:
        while process.is_alive():
            time.sleep(0.1)
        globals.server.should_exit = True
        while globals.state != globals.State.STOPPED:
            time.sleep(0.1)
        _thread.interrupt_main()

    if 'native' not in globals.optional_features:
        globals.log.error('Native mode is not supported in this configuration.\n'
                          'Please run "pip install pywebview" to use it.')
        sys.exit(1)
 
    mp.freeze_support()
    args = host, port, title, width, height, fullscreen, frameless, native.method_queue, native.response_queue
    process = mp.Process(target=open_window, args=args, kwargs=globals.app.native.window_args, daemon=False) # <- add kwargs directly to the Process call
    process.start()

    Thread(target=check_shutdown, daemon=True).start()
```
works and the app.native.window_args are carried over as expected. Thus it seems to be a multiprocessing issue even more. Moreover the  
```python
 mp.freeze_support()
```
might be out of place, at least if i read the multiprocessing docs correctly (?!):
```
multiprocessing.freeze_support()
Add support for when a program which uses [multiprocessing](https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing) has been frozen to produce a Windows executable. (Has been tested with py2exe, PyInstaller and cx_Freeze.)

One needs to call this function straight after the if __name__ == '__main__' line of the main module.
```
Maybe someone with more experience can make some sense out of it. 

@rodja Are you 100% sure the way the *nicegui.globals* module is used works for the native mode window args? 
I cannot test with *pyinstaller* on a windows machine at the moment, but maybe anyone already has?

**Comment by rodja** [comment]
>  `mp.freeze_support()` might be out of place

Yes, from the documentation it looks wired. I remember vaguely that it still was necessary to ensure the window is closing properly when I worked on https://github.com/zauberzeug/nicegui/pull/917.

> Are you 100% sure the way the nicegui.globals module is used works for the native mode window args?

Yes. `native_mode.activate` is called from the NiceGUI main thread which has access to `globals.app.native.window_args` and then get passed to the `mp.Process` which starts the pywebview window. I've also just tested your example which worked fine (after I added `reload=False` as mentioned in the docs):

```py
from nicegui import app, ui

ui.label('Hello Frameless World!')
app.native.window_args['frameless'] = True
app.native.window_args['confirm_close'] = True

ui.run(native=True, reload=False)
```

**Reply by FabianGoessling** [reply]
> Are you 100% sure the way the nicegui.globals module is used works for the native mode window args?

I was wondering about the mp.Process() because i can confirm the *activate* function has access, but the start_window() inside the new process gets empty i.e. the default "globals". Seems like there is a completely "fresh" init happening.. 

So i understand correctly you tried your example in "frozen" mode, i.e. pyinstaller? So i have to dig into *shiv* or just switch to pyinstaller 😂..

**Comment by chaosct** [comment]
Hi there,

I am being impacted by this very issue but without the need of compiling or freezing the code (just running with `uv run`). I made a pull request that solves the issue for me here: https://github.com/zauberzeug/nicegui/pull/4665

**Reply by FabianGoessling** [reply]
Hi, thanks, i did not look further and instead used my monkeypatch since. Will take a look if you PR solves my problem.

**Reply by chaosct** [reply]
Hi! In fact, I was having two separate issues. I am now having your exact issue with pyinstaller.

Can you share your monkeypatch solution? I'd prefer not to use a forked version of nicegui for production...

**Reply by FabianGoessling** [reply]
Sorry for the late answer. My monkeypatch is basically what I posted in above - I rewrote the open_window function in my project and then I did a monkeypatch using pytest at runtime 🤣
The problem is that my solution will probably not work out of the box for a newer nicegui version.

**Comment by FabianGoessling** [comment]
More than 2 years later I solved the problem and this might be a solution for some of the other native mode in frozen/packaged apps as well. Note that from my perspective this to nothing to do with a wrong main guard or placement of the app.native.window_args. 

Setup:
```python
from nicegui import ui, app

app.native.window_args['frameless'] = True #--> does not work
app.native.window_args['confirm_close'] = True #--> does not work
app.on_startup(do_some())

if __name__ == "__main__": # Can be in ["__main__", "__mp_main__"]
       ui.run(frameless=True) # --> works
```

Here the frameless=True parameter is passed directly (and it works), while the app.native_args are not respected if the app is started packaged (in my case by shiv). The reason is how the multiprocessing spawns its child process, which is by looking at the `__spec__` of the parents `__main__` file. Now if we simply start by

```python
python -m my_script 
```

my_script is the main and will be spawned (and rerun) as mp_main as well. For the case of shiv, there is an additional bootstrap file and the startup of 

```python
python my_libs.pyz -m my_script 
```

does in fact run another **main** file embedded in the my_libs.pyz and sets the `__spec__` to that main file. I.e. the multiprocessing reruns the bootstrap file instead of *my_script* which then cannot apply the app.native.window_args. 

This additionally explains why the monkeypatch above of directly replacing the _open_window function worked. Now the solution is to add the following lines in the *my_script* script, which should not harm the unpackaged startup:

```python
import sys 
if sys.modules["__main__"].__spec__:
         sys.modules["__main__"].__spec__.origin = __file__
         sys.modules["__main__"].__spec__.name = "replace_this_with_the_scripts_name"
```

If this does not make any sense, let's blame it on Claude, but I am happy cause it solved my problem (and this is connected to setting the icon under #1745 as well).

Looking forward to hearing ideas @falkoschindler, @evnchn, @chaosct, @J3ronimo, this could be helpful to solve some of these #5648, #4842, #4665 or #4627. If not, at least the problem is solved for shiv.

**Reply by falkoschindler** [reply]
Thanks for the thorough analysis, @FabianGoessling! This is really helpful.

We're considering whether NiceGUI could detect and fix this automatically in `activate()` before spawning the subprocess. When `__spec__.origin` doesn't match `__file__`, multiprocessing respawns the wrong file. Rather than trying to fix both `__spec__.origin` and `__spec__.name` (deriving the correct module name from NiceGUI's side would be tricky), we could set `__spec__` to `None` entirely — that makes multiprocessing fall back to path-based re-execution using `__main__.__file__`:

```python
main = sys.modules['__main__']
if hasattr(main, '__spec__') and main.__spec__ and main.__spec__.origin != getattr(main, '__file__', None):
    main.__spec__ = None
```

Could you check what `sys.modules['__main__'].__file__` and `sys.modules['__main__'].__spec__.origin` look like in your shiv case? If `__file__` reliably points to your actual script while `__spec__.origin` points to the bootstrap, this approach could work. And does setting `__spec__ = None` break anything in your setup?
